M2/Macaulay2/packages/MonomialIntegerPrograms.m2

```diff
@@ -0,0 +1,1656 @@
+  codimensionIP
+  MonomialIntegerPrograms
+  symbol ScipPrintLevel
+///
+
+doc ///
+ Key
+  codimensionIP
+  (codimensionIP, MonomialIdeal)
+ Headline
+  compute the codimension of a monomial ideal using integer programming
+ Usage
+  c = codimensionIP(I)
+ Inputs
+  I:MonomialIdeal
+ Outputs
+  c:ZZ
+   the codimension of $I$
+ Description
+  Text
+   The integer programming input and output files created are named
+   "codim.zpl", "codim.errors", etc., and saved to a temporary directory.
+   By default the location of the temporary directory is printed to the
+   screen.
+  Example
+   R = QQ[x,y,z,w,v];
+   I = monomialIdeal(x*y*w, x*z*v, y*x, y*z*v);
+   codimensionIP(I)
+  Text
+   The verbosity of every function in the MonomialIntegerPrograms package is controlled with
+   @TO symbol ScipPrintLevel@. For example, to suppress printing the name of
+   the directory or any other information and simply return the answer, set
+   @TT"ScipPrintLevel"@ to 0.
+  Example
+   ScipPrintLevel = 0;
+   J = monomialIdeal(x*y^3*z^7, y^4*w*v, z^2*v^8, x*w^3*v^3, y^10, z^10)
+   codimensionIP(J) 
+  Text
+   The codimension of a monomial ideal is equal to the codimension
+   of its radical. Therefore, when looking at the IP formulation written to
+   the temporary file "codim.zpl", you will see that exponents are ignored.
+ SeeAlso
+  (codim, MonomialIdeal)
+  dimensionIP
+  MonomialIntegerPrograms
+  symbol ScipPrintLevel
+///
+
+doc ///
+ Key
+  monomialIdealsWithHilbertFunction
+  (monomialIdealsWithHilbertFunction, List, PolynomialRing)
+  [monomialIdealsWithHilbertFunction, BoundGenerators]
+  [monomialIdealsWithHilbertFunction, FirstBetti]
+  [monomialIdealsWithHilbertFunction, GradedBettis]
+  [monomialIdealsWithHilbertFunction, SquareFree]
+ Headline
+  find all monomial ideals in a polynomial ring with a particular (partial or complete) Hilbert function
+ Usage
+  monomialIdealsWithHilbertFunction(L, R)
+  monomialIdealsWithHilbertFunction(L, R, BoundGenerators => a)
+  monomialIdealsWithHilbertFunction(L, R, FirstBetti => b)
+  monomialIdealsWithHilbertFunction(L, R, GradedBettis => B)
+  monomialIdealsWithHilbertFunction(L, R, SquareFree => x)
+ Inputs
+  L: List
+   $\{h(0), h(1), \ldots, h(d)\}$, the values of a valid Hilbert function for $R$ for degrees $0\ldots d$.
+  R: PolynomialRing
+  BoundGenerators => ZZ
+   a degree bound on the monomial generators
+  FirstBetti => ZZ
+   a specified first total Betti number
+  GradedBettis => List
+   if the form $\{b_0, b_1, \ldots, b_d\}$, where $b_i$ is the first graded Betti numbers with degree $i$.
+  SquareFree => Boolean
+   whether or not to consider squarefree monomial ideals only
+ Outputs
+   :List
+    all ideals $I$ of $R$ that satisfy $HF(R/I, i) = h(i)$ for all $0\le i\le d$
+ Description
+  Text
+   For example, count the monomial ideals in $\mathbb{Q}[x,y,z]$, generated in degrees up to 5, whose Hilbert
+   function begins with $\{1, 3, 6, 5, 4, 4\}$:
+  CannedExample
```

Comment:
Why use any canned examples?  To test the code works, it would be better if it were an ordinary example, so it would get rerun from time to time.